Add functionality to graph DHCP stats on each VLAN page and Prefix page in NAV

jorund1 opened the pull request:
## Scope and purpose

Fixes #2373. Another PR intends to add documentation and release notes.

### This pull request

* Adds DHCP stats graphs to VLAN and Prefix pages in NAV where recent-enough stats for that VLAN/Prefix are found in the Carbon/Graphite database:
  ![dhcp-frontend-delimited-timeseries-showcase](https://github.com/user-attachments/assets/19bbff15-ed78-4c36-8eca-2615cdd1511a)

## Contributor Checklist

Every pull request should have this checklist filled out, no matter how small it is.
More information about contributing to NAV can be found in the
[Hacker's guide to NAV](https://nav.readthedocs.io/en/latest/hacking/hacking.html#hacker-s-guide-to-nav).

<!-- Add an "X" inside the brackets to confirm -->
<!-- If not checking one or more of the boxes, please explain why below each or remove the line if not applicable. -->

* [x] Added a changelog fragment for [towncrier](https://nav.readthedocs.io/en/latest/hacking/hacking.html#adding-a-changelog-entry)
* [x] Added/amended tests for new/changed code
* [ ] ~Added/changed documentation~ see #3785 
* [x] Linted/formatted the code with ruff, easiest by using [pre-commit](https://nav.readthedocs.io/en/latest/hacking/hacking.html#pre-commit-hooks-and-ruff)
* [x] Wrote the commit message so that the first line continues the sentence "If applied, this commit will ...", starts with a capital letter, does not end with punctuation and is 50 characters or less long. See https://cbea.ms/git-commit/
* [x] Based this pull request on the correct upstream branch: For a patch/bugfix affecting the latest stable version, it should be based on that version's branch (`<major>.<minor>.x`). For a new feature or other additions, it should be based on `master`.
* [x] If applicable: Created new issues if this PR does not fix the issue completely/there is further work to be done **TODO: add a new issue for using Rickshaw graphs instead of PNG images in graph navlets**
* [x] If it's not obvious from a linked issue, describe how to interact with NAV in order for a reviewer to observe the effects of this change first-hand (commands, URLs, UI interactions) **See screencast below**
* [x] If this results in changes in the UI: Added screenshots of the before and after
* [x] If this adds a new Python source code file: Added the [boilerplate header](https://nav.readthedocs.io/en/latest/hacking/hacking.html#python-boilerplate-headers) to that file


## Testing It For Yourself

You can use the script below to fill the prefixes  `192.0.0.0/24`, `192.0.1.0/24`, `192.0.0.0/26`, and `::2:0:0/96` with DHCP stats (you might need to install `socat` and `gawk`, and you might need to change `127.0.0.1` to the address of your Carbon/Graphite database):

```bash
#!/bin/sh

cat <<EOF |
4 example-1 admin 192.0.0.1 192.0.0.127 127 127 0 100000
4 example-1 admin 192.0.0.129 192.0.0.255 127 127 0 100000
4 example-1 admin 192.0.1.1 192.0.1.255 255 255 0 100000

4 example-1 guest 192.0.2.0 192.0.2.255 256 256 0 10000
4 example-1 guest 192.0.2.0 192.0.2.127 128 128 10001 50000
4 example-1 guest 192.0.2.128 192.0.2.191 64 64 10001 50000
4 example-1 guest 192.0.2.192 192.0.2.255 64 64 10001 50000

6 example-1 admin 0:0:0:0:0:2:0:0 0:0:0:0:0:2:ffff:ffff 4294967296 100 0 100000
6 example-1 guest 0:0:0:0:0:2:0:0 0:0:0:0:0:2:ffff:ffff 4294967296 100 0 100000
EOF

gawk '
BEGIN {
  NOW = systime()
  srand(11)
  INTERVAL = 300
  DAYS = 30
}

NF == 9 {
  a = rand()
  b = rand()
  c = rand()
  _ = 1 / (a + b + c)
  a = a * _
  b = b * _
  c = c * _
  d = rand()*3.14*2
  e = 20 + rand()*8
  f = rand()

  ip_version = $1
  server_name = $2
  group_name = $3
  first_ip = $4; gsub(/[.:]/, "_", first_ip)
  last_ip = $5; gsub(/[.:]/, "_", last_ip)
  total_ips = $6
  upper_limit = $7
  start_offset = $8 * INTERVAL
  end_offset = $9 * INTERVAL
  noise = rand()
  for (offset = start_offset; offset < end_offset; offset += INTERVAL) {
    sine = (sin((offset*3.14*2)/(3600*e) + d)+1)/2
    noise = noise*f + rand()*(1-f)
    assigned_ips = int(sine*upper_limit*a+upper_limit*b+noise*upper_limit*c)
    printf("nav.dhcp.servers.%s.range.custom_groups.%s.%d.%s.%s.unassigned %d %d\n", server_name, group_name, ip_version, first_ip, last_ip, total_ips - assigned_ips, NOW - offset)
    printf("nav.dhcp.servers.%s.range.custom_groups.%s.%d.%s.%s.assigned %d %d\n", server_name, group_name, ip_version, first_ip, last_ip, assigned_ips, NOW - offset)
    printf("nav.dhcp.servers.%s.range.custom_groups.%s.%d.%s.%s.total %d %d\n", server_name, group_name, ip_version, first_ip, last_ip, total_ips, NOW - offset)
  }
}' |

socat - tcp:127.0.0.1:2003
```

Here's a screencast (which, after the most recent changes, is a bit outdated, but still representative):

https://github.com/user-attachments/assets/57b7ccf9-4842-4e88-8e0d-fa3cd65de5b4

At `2:00` in the video, notice that the graph labelled `DHCP ranges in 'guest' on server 'kea-trd'` has stats from ranges in both `192.0.2.0/24` and `198.51.100.0/24` despite vlan 20 only containing `192.0.2.0/24`. This is because whenever a group (here: `'guest' on server 'kea-trd'`) has at least one range/pool/subnet overlapping with the VLAN/Prefix, the whole group will be displayed on that VLAN/Prefix's page.

jorund1 committed:
Add DHCP graphs to the VLAN and Prefix views in NAV

..a VLAN/Prefix page has DHCP graphs displayed if NAV has collected
some recent enough DHCP stats for IP addresses that intersect that
VLAN/Prefix.

jorund1 committed:
Add changelog entry

lunkwill42 commented on python/nav/dhcpstats/common.py:
```diff
@@ -0,0 +1,216 @@
+            group_name = group_name
+
+        first_ip = IPy.IP(IPy.IP(first_ip)[0])
+        last_ip = IPy.IP(IPy.IP(last_ip)[-1])
```
This should be unnecessary. Inputting prefixes into the `first_ip` and `last_ip` arguments should be considered an error, rather than trying to do some weird guesstimation and allow it...

lunkwill42 replied:
I'd say this is in the blocking category (these two lines just look weird)

lunkwill42 commented on python/nav/dhcpstats/common.py:
```diff
@@ -0,0 +1,216 @@
+                or self.first_ip < prefix < self.last_ip
+            ):
+                return True
+        return False
```
Could potentially use `any()` here.

lunkwill42 replied:
This is non-blocking, just a suggestion

jorund1 replied:
fixed

lunkwill42 reviewed (changes requested):
Overall the code seems nicely structured. After a long F2F session, I would say my most important suggestion is to split this into two PRs: One for the rewrite of the backend functionality, and one for the purely front-end related additions.

Lastly, I understand there is also a documentation PR coming, which is nice.

lunkwill42 commented:
> Overall the code seems nicely structured. After a long F2F session, I would say my most important suggestion is to split this into two PRs: One for the rewrite of the backend functionality, and one for the purely front-end related additions.
> 
> Lastly, I understand there is also a documentation PR coming, which is nice.

Also a reminder to properly enable the pre-commit hooks in each clone you have: Linting failures abound here...

jorund1 committed:
Add helper functions for rendering Graphite graphs

These can be used instead of string interpolations which quickly become
unwieldy when you want a little bit more advanced Graphite renders.

jorund1 committed:
Add functionality for making DHCP graphs for a set of prefixes

The function fetch_graph_urls_for_prefixes takes a set of prefixes
(e.g. obtained from a models.manage.Prefix or from a models.manage.Vlan) and
returns one url per DHCP graph from Graphite related to one or more of the
prefixes. Each url returns a JSON with graph data.

lunkwill42 reviewed (approved):
Alright, let's go with this!

I have an inline concern that should be followed up with a new issue or PR.

There's also this non-blocking follow-up item: If Graphite is unavailable, I get a very useful red "Graphite unavailable" message in the GUI. However, the backend request fails with an unhandled 500 message, which ultimately gets sent as an e-mail to the site admin.  These should really be handled, there is no reason to bother the site admin with lots of tracebacks for an issue that ultimately is not a problem in NAV but a problem with how Graphite is operating (which should be monitored in other ways)
